### PR TITLE
Feat/17 api gateway

### PR DESCRIPTION
**Pull request by simbayippy**
### to test
docker compose up to run all other services
cd into api-gateway and pnpm dev

### about
API Gateway now allows it such that the client side only needs to call 1 api host (the api gateway itself) instead of each individual microservice.

now to communicate with the microservices from the FE, we just need to use a common axiosClient instance

API Gateway will then route the requests accordingly based on pattern matching to the individual microservices

image

### examples
axiosClient found in @network/axiosClient.ts

```
const axiosClient = axios.create({
  baseURL: `${API_GATEWAY_URL}/api`,
  headers: {
    'Content-Type': 'application/json',
  },
});
```

Example usage:

```
const response = await axiosClient.get('/questions/tags');
```

This routes the request through the api gateway which will then be handled by the questionRouter.ts (found in /api-gateway/routers)

### CODE REVIEW

File: api-gateway/src/app.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+app.use(express.json());
+
+// Logging middleware
+app.use((req, res, next) => {
+  logger.info(`Incoming request: ${req.method} ${req.originalUrl}`);
+  next();
+});
+
+// Routes
+app.use('/api', routes);
```
Comment (Kurtyjlee) [inline api-gateway/src/app.ts:24]: can we add v1 to this? just looks nicer
```suggestion
app.use('/api/v1', routes);
```

File: api-gateway/src/routes/authRoutes.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  changeOrigin: true,
+  pathRewrite: {
+    '^/(.*)': '/api/v1/auth/$1', // Maps /login -> /api/v1/auth/login
+  },
+  on: {
+    proxyReq: fixRequestBody, // need to fix the request body before forwarding
+  },
+});
+
+// Log incoming requests to the Auth route
```
Comment (Kurtyjlee) [inline api-gateway/src/routes/authRoutes.ts:20]: Are we meant to use the logger here? If the logger is important then maybe we want to enable it

File: api-gateway/src/routes/matchingRoutes.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { authMiddleware } from '../middleware/authMiddleware';
+
+const router = express.Router();
+
+const matchingServiceProxy = createProxyMiddleware({
+  target: config.services.matching, // Matching Service URL
+  changeOrigin: true,
+  ws: true, // Enable WebSocket proxying
+  pathRewrite: {
+    '^/(.*)': '/$1', // front end connects to matching service via /ws only
```
Comment (Kurtyjlee) [inline api-gateway/src/routes/matchingRoutes.ts:14]: might want to still add the `api/v1/` here incase we want to do a health ping? Currently `match-svc` only has 1 route for health check

File: api-gateway/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+COPY pnpm-lock.yaml ./
+
+RUN pnpm install
+
+# Development stage
+FROM base AS development
+
+COPY src ./src
+COPY tsconfig.json ./
+COPY .env.dev ./
```
Comment (Kurtyjlee) [inline api-gateway/Dockerfile:21]: I think we dont need to copy in the .env.dev as it is referenced in the compose.yml

File: peerprep-fe/src/network/axiosClient.ts [hunk 5]
```diff
@@ -1,45 +1,18 @@
...
-    'Content-Type': 'application/json',
-  },
-});
+const API_GATEWAY_URL =
+  process.env.NEXT_PUBLIC_API_GATEWAY_URL || 'http://localhost:8001';
 
-const axiosAuthClient = axios.create({
-  baseURL:
-    process.env.NEXT_PUBLIC_AUTH_SERVICE_URL || 'http://localhost:3001/api/v1',
+const axiosClient = axios.create({
```
Comment (Kurtyjlee) [inline peerprep-fe/src/network/axiosClient.ts:7]: Nice this looks dam clean now

**Review by Kurtyjlee — COMMENTED** [review]
Just some nits, lmk what you think :))

File: api-gateway/src/routes/authRoutes.ts (hunk 2 above)
Comment (simbayippy) [reply api-gateway/src/routes/authRoutes.ts:20]: i think its ok, if we enable logging it'll flood the terminal bc itll be a ton of incoming messages

**Review by tyouwei — APPROVED** [review]
LGTM!!!

**Pull request merged**